[REV] web_editor: revert fix for losing text-gradient class on preview

sobo-odoo opened the pull request:
This PR reverts [this fix](https://github.com/odoo/odoo/commit/994fe756db8f1a45ef864cb7848644ea204539b7), which was made to avoid losing the `text-gradient` class on preview, because it broke other behaviors, such as not saving the mega menu when removing mega menu items (because the `o_dirty` class was not added anymore).
This PR then fixes the `text-gradient` class issue in an other way.
The last commit fixes an issue happening when undoing a column count change. This issue was indirectly fixed by the reverted commit as mentioned [here](https://github.com/odoo/odoo/pull/118872#issuecomment-1554414305), but now that it was reverted, the issue is back.

See commit messages for more details.

opw-3630140

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/147495).

sobo-odoo commented:
@qsm-odoo @rdeodoo ready for review :slightly_smiling_face: 

As a fix, I reverted the problematic commit and fixed what it was fixing in an other way :slightly_smiling_face: 
The third commit was mentioned in the [original fix PR](https://github.com/odoo/odoo/pull/118872#issuecomment-1554414305) so I fixed it too -> I don't know what to do with its canceled [corresponding task](https://www.odoo.com/web#id=3330310&menu_id=4720&cids=1&action=333&active_id=974&model=project.task&view_type=form) though, should I reopen it and add my PR in it ? 
Same question for the [text gradient task](https://www.odoo.com/web#id=3120090&menu_id=4720&cids=1&action=333&active_id=974&model=project.task&view_type=form) :slightly_smiling_face: 

I think a better fix should rather filter better the cases where a class is removed and then added back (meaning that no change happened in the end concerning this class), something like cancelling opposite mutations or merging them
Maybe something like this:
- Get all class mutations (attributeName = class)
- by target, look at the added and removed classes:
  - if removed (added), add it in a set, and when the opposite is done, we remove it from the set -> in the end, only the relevant classes will be in the set
- Only make one class mutation that contains the global class changes (we could store a "removed"/"added" information with the class in the set -> the removed ones will be in `oldValue` and the added will already be on the target

Not sure if it is a good idea or even if its is worth it (we have easy ways to avoid this issue after all -> make the changes in one step) but in any case, it would be for the editor team anyway :thinking:

sobo-odoo commented:
@qsm-odoo @rdeodoo repinging :slightly_smiling_face:  (sorry for the spam, but it is for a ticket)

qsm-odoo commented on addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:
```diff
@@ -658,7 +658,7 @@ export class OdooEditor extends EventTarget {
                         }
                     }
-                    if (excludedClasses.length && excludedClasses.every(c => this.options.renderingClasses.includes(c))) {
+                    if (excludedClasses.every(c => this.options.renderingClasses.includes(c))) {
```
`excludedClasses` seems to be the most misleading name for a variable I have ever seen, does it actually mean "classes that were added or removed of the element"? :grimacing: (if yes we can probably change that directly here in stable, it is just a local variable and I am sure I will be debugging this code over the next 3 years so...).

And if this is indeed that, maybe comment the whole section:
```javascript
// First list the classes that were actually removed/added from the element
// ...
// Then ignore the mutation if ...
```
This would actually help me understand this because I don't: the commit you are reverting introduces a test that says "skip the mutations if no changes in state". If I am not confused, "no changes in state" means `excludedClasses.length === 0`... but before their commit we just did not check the `.length` and thus call `.every(...)` directly which means `true`, which means we `continue`... and thus skip the mutation already... no? And with their fix we check the `.length` thus it is falsy since no change of state, so we do not `continue` here and thus consider the mutation? I am very confused: it feels like their fix is doing the opposite of what they meant but a test ensures otherwise :grimacing:

qsm-odoo commented on addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:
```diff
@@ -237,30 +237,32 @@ function align(editor, mode) {
         .replace(mode === 'color' ? TEXT_CLASSES_REGEX : BG_CLASSES_REGEX, '')
         .replace(/\btext-gradient\b/g, '') // cannot be combined with setting a background
         .replace(/\s+/, ' ');
-    element.className !== newClassName && (element.className = newClassName);
```
What the hell was this, respect is really dead -_-

qsm-odoo commented on addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:
```diff
@@ -237,30 +237,32 @@ function align(editor, mode) {
     } else {
         element.style[mode] = color;
     }
+    // Replacing the classes only once at the end to have only one class
+    // mutation in the current history step.
+    element.className = newClassName;
```
I don't think we should fix the issue like that (although I am very tempted to remove the awful line above). We should not have to think about the mutations an edition code will trigger.

I am not sure what the solution is though, I don't understand your paragraph in the previous commit:
> - When we stop hovering (`color_leave` event), we revert the current
> step and so the mutation records are filtered first. For classes, there
> will be filtered this way (the element currently has the `text-gradient`
> class):
>   - 1st mutation: the oldValue is in the element classes and vice versa
>    -> no `excludedClasses`
>       => the record is discarded (`continue`)
>   - 2nd mutation: the current element class is not in oldValue
>     -> `text-gradient` is added in `excludedClasses` and it is not in
>    `renderingClasses` so the condition is false
>       => the record is kept

What does "we revert the current step and so the mutation records are filtered first. " actually means? We don't call colorElement back? We relying on the previously recorded mutations? If that is so, why are filtering them again? (Maybe I don't understand anything about the way it works here, you tell me ^^).

Also I would squash both commits. It makes sense for a commit to mention its previous one (and I think here you did as RDE suggested in case of a "revert" but RDE answered generically, without knowing what this was actually about I think) but in this case:
- Your revert reintroduces a bug (at least I think?) ->  we should not reintroduce bugs, either fix them the right way first and then revert the other problematic commit which created other bugs, or:
- In this case, I would squash because I cannot understand this commit without the previous one (I always had to jump back and forth between your previous commit message and this diff here to try and understand).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4788,9 +4788,9 @@ registry.layout_column = SnippetOptionWidget.extend({
-            var $column = $(column);
-            $column.attr('class', $column.attr('class').replace(/\b(col|offset)-lg(-\d+)?\b/g, ''));
-            $column.addClass(colClass);
+            let className = column.className.replace(/\b(col|offset)-lg(-\d+)?\b/g, "");
+            className += (className ? " " : "") + colClass;
+            column.className = className;
```
Same remark: we should not have to care about this (well, we can to make more pretty code, but the current implementation should work). I would be very surprised if there were actually only two occurences of some code doing that... and if you consider community apps & custos... :grimacing:

qsm-odoo reviewed (changes requested):
Thanks for the investigations and explanations, but I think we need to do more :) Also sorry for the delay, I saw all your pings but my current TODO list is big :wink: 

Also maybe clarify in your revert this:
> Since commit [1], removing a mega menu item is not saved anymore because
the `o_dirty` class is not added, making the changes being ignored when
saving.

=> I cannot reproduce that bug, maybe need clearer steps to reproduce. Note that I saw in the OPW that it could be related to the o_default_snippet_text class which was removed on click... but I fixed that last week with https://github.com/odoo/odoo/commit/e2cd6940bd29670ea0e76cd4bb95489a12588f55 (which I have to fix even more actually).

sobo-odoo commented:
> => I cannot reproduce that bug, maybe need clearer steps to reproduce. Note that I saw in the OPW that it could be related to the o_default_snippet_text class which was removed on click... but I fixed that last week with https://github.com/odoo/odoo/commit/e2cd6940bd29670ea0e76cd4bb95489a12588f55 (which I have to fix even more actually).

@qsm-odoo Well, it apparently doesn't happen anymore since the [last commit](https://github.com/odoo/odoo/commit/9b26aafbc1576b08c409344d22ec179bd32a3bd7) of the PR your commit comes from... My bad I should have rechecked before spamming :grimacing: (I didn't add steps because it was really just deleting a mega menu link, but now I realize that I wrote "item", which is confusing indeed :grimacing: )

But the fix I did was based on the commit that introduced the issue so I thought something was wrong with it and when looking at what it does and why it was needed, I found it a bit weird (I will explain better when answering your other comment :slightly_smiling_face: ) Also because since it had an impact on the `o_dirty` class, I found it suspicious to break more than it fixed

Note that with your fixes, the `o_dirty` class is still added on the mega menu when clicking for the very first time on one of its link, even though nothing changed (it is not the case when we click after having saved). Just telling so you know, since you said you will fix even more :slightly_smiling_face:

qsm-odoo commented:
@sobo-odoo 

> it apparently doesn't happen anymore since the [last commit](https://github.com/odoo/odoo/commit/9b26aafbc1576b08c409344d22ec179bd32a3bd7) of the PR your commit comes from

Ah funny, don't know why, don't need to know why :smile: 

> My bad I should have rechecked before spamming

Of course not, no need to recheck every week that your PR still does what is is supposed to do :wink:  If it was fixed some other way in the mean time, that's my fault for not checking your PR sooner and if the diff still makes sense despite the original bug being fixed, it's still worth checking.

> Note that with your fixes, the o_dirty class is still added on the mega menu when clicking for the very first time on one of its link

Someone will have to check then, no need to create a task, I will do it with a list of some more "dirty for reason" problems :grimacing:

sobo-odoo commented:
@qsm-odoo 

From my previous answer
> ... I found it a bit weird (I will explain better when answering your other comment 🙂 ) Also because since it had an impact on the o_dirty class, I found it suspicious to break more than it fixed

When trying to explain it to answer [your comment](https://github.com/odoo/odoo/pull/147495#discussion_r1448626971), I realized that in fact their fix is alright :grimacing: (I didn't pay much attention at the code around it... :grimacing: )

So to answer what you asked:

> excludedClasses seems to be the most misleading name for a variable I have ever seen, does it actually mean "classes that were added or removed of the element"? 

=> yes, it is the classes that changed in this mutation. I guess they called it `excludedClasses` to "exclude" them from the `renderingClasses` :shrug: 

> If I am not confused, "no changes in state" means excludedClasses.length === 0... but before their commit we just did not check the .length and thus call .every(...) directly which means true, which means we continue... and thus skip the mutation already... no? And with their fix we check the .length thus it is falsy since no change of state, so we do not continue here and thus consider the mutation? I am very confused: it feels like their fix is doing the opposite of what they meant but a test ensures otherwise

- Before their fix:
  - If all the classes that changed are in the list of the classes to skip, we skip the mutation (-> continue). 
  - If nothing changed, then also skip the mutation (which makes sense, since nothing changed) 
- After their fix: 
  - If some classes changed and if they are all to skip, we skip the mutation.
  - If nothing changed, we don't skip.

 => We don't skip because we still need to store the first attribute change in `attributeCache` (so we only consider the very first mutation for a given attribute type) 
So, with their fix, we have: 
classBefore: [`text-gradient`], classAfter: [`text-gradient`] => `excludedClasses` is empty => do not skip => save this step as the first class mutation (= mark it in `attributeCache`) -> when reading the code, we see that in the cache for this element, the `class` attribute is set to `false`
 => ignore all the class mutations (for this element): since `class` is set to `false`, it goes to `continue` and so the records are not stored.

That's why it fixed the gradient issue, because no class mutations needed to be reverted :slightly_smiling_face: 

So in the end, we can close this PR (and its ticket :grimacing:) if you agree :slightly_smiling_face: (unless you still want to rename `excludedClasses` haha) Since the mega menu issue doesn't happen anymore (with your fix about the links, see my previous comment) and since their fix was not as weird as I thought at first :grimacing:

qsm-odoo commented:
@sobo-odoo I did not understand it all but I don't need to. If you think nothing is to be fixed anymore, I trust you :) You can then close the PR :+1: Do you prefer I send a message to the customer or will you do it? :+1:

sobo-odoo commented:
> Do you prefer I send a message to the customer or will you do it?

I can do it :slightly_smiling_face: Thanks :+1:
